src/termination/ReachabilityNonterm.cc

```diff
@@ -0,0 +1,236 @@
+    chcs.addClause(ChcHead{UninterpretedPredicate{logic.getTerm_false()}},
+                   ChcBody{InterpretedFla{query}, {UninterpretedPredicate{pred}}});
+    Normalizer normalizer(logic);
+    auto normalizedSystem = normalizer.normalize(chcs);
+    auto hypergraph = ChcGraphBuilder(logic).buildGraph(normalizedSystem);
+    assert(hypergraph->isNormalGraph());
+    return hypergraph;
+}
+
+ReachabilityNonterm::Answer ReachabilityNonterm::nontermination(TransitionSystem const & ts) {
+    auto vars = ts.getStateVars();
+    ArithLogic & logic = dynamic_cast<ArithLogic &>(ts.getLogic());
+    PTRef init = ts.getInit();
+    PTRef transition = ts.getTransition();
+    // In this case query is a set of sink states - states from which transition is not possible.
+    // sink /\ transition is UNSAT
+    PTRef sink = logic.mkNot(QuantifierElimination(logic).keepOnly(transition, vars));
+
+    // if sink is false, there are no sink states in the TS, therefore it is nonterminating
+    if (sink == logic.getTerm_false()) {
+        std::cout << "; (Trivial nontermination)" << std::endl;
+        return Answer::NO;
+    }
+
+    // Witness computation is required, as we need to use both counterexample traces to limit terminating states
+    // and inductive invariants to prove nontermination
+    Options witnesses = options;
+    witnesses.addOption(options.COMPUTE_WITNESS, "true");
+
+    // Main nonterm-checking loop
+    while (true) {
+        // Constructing a graph based on the currently considered TS
+        auto graph = constructHyperGraph(init, transition, sink, logic, vars);
+        auto engine = EngineFactory(logic, witnesses).getEngine(witnesses.getOrDefault(Options::ENGINE, "spacer"));
+
+        // Check if sink states are reachable within TS
+        auto res = engine->solve(*graph);
+        if (res.getAnswer() == VerificationAnswer::UNSAFE) {
+            // When sink states are reachable, extract the number of transitions needed to reach the sink states
+            uint num = res.getInvalidityWitness().getDerivation().size() - 3;
+
+            // Construct the logical formula representing the trace:
+            // Init(x) /\ Tr(x,x') /\ ... /\ Bad(x^(num))
+            std::vector formulas{init, TimeMachine(logic).sendFlaThroughTime(sink, num)};
+            for (uint j = 0; j < num; j++) {
+                formulas.push_back(TimeMachine(logic).sendFlaThroughTime(transition, j));
+            }
+            PTRef transitions = logic.mkAnd(formulas);
+
+            // Get the satisfying model of the trace.
+            // It is needed to detect nondeterminism
+            SMTSolver SMTsolver(logic, SMTSolver::WitnessProduction::ONLY_MODEL);
+            SMTsolver.assertProp(transitions);
+            if (SMTsolver.check() != SMTSolver::Answer::SAT) {
+                assert(false);
+                return Answer::ERROR;
+            }
+            auto model = SMTsolver.getModel();
```

Comment:
I think you can get the same information directly from the invalidity witness.
But maybe for a prototype, this is also OK.

Reply:
Hmm, how though, the only function in the `InvalidityVitness` to get information is `getDerivation`, and there is no function in the `Derivation` to get a model or a PTRef...